File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = -1
    graph = [[] for _ in range((n+1))]
    
    for a,b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def bfs(start):
        visit = [0]*(n+1)
        que = deque([start])
        visit[start] =1
        cnt=1
        while que:
            s = que.popleft()
            for i in graph[s]:
                if not visit[i]:
                    que.append(i)
                    visit[i] = 1
                    cnt+=1
        return cnt     
    res = 100000000
    for a,b in wires:
        graph[a].remove(b)
        graph[b].remove(a)
        res = min(res, abs(bfs(a) - bfs(b)))
        graph[a].append(b)
        graph[b].append(a)
    return res

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    static ArrayList<Integer>[] arr;
    
    public int solution(int n, int[][] wires) {
        int answer = 1_000_000_000;
        arr = new ArrayList[n+1];
        for(int i=1;i<=n;i++){
            arr[i] = new ArrayList<>();
        }

        
        for(int i=0;i<wires.length;i++){
            int[] tmp = wires[i];
            arr[tmp[0]].add(tmp[1]);
            arr[tmp[1]].add(tmp[0]);
        }
        
        for(int i=0;i<wires.length;i++){
            int a= wires[i][0];
            int b = wires[i][1];
            
            boolean[] visit = new boolean[n+1];
            
            arr[a].remove(Integer.valueOf(b));
            arr[b].remove(Integer.valueOf(a));
            
            int cnt1 = bfs(visit,a);
            int cnt2 = bfs(visit,b);
            
            answer = Math.min(answer, Math.abs(cnt1-cnt2));
            
            arr[a].add(b);
            arr[b].add(a);
        }
        
        return answer;
    }
    
    public static int bfs(boolean[] visit, int start){
        Queue<Integer> que = new LinkedList<>();
        que.add(start);
        visit[start] = true;
        int cnt=0;
        while(!que.isEmpty()){
            int tmp = que.poll();
            cnt++;
            for(int t:arr[tmp]){
                if(!visit[t]){
                    que.add(t);
                    visit[t] = true;
                }
            }
        }
        return cnt;
    }
    
}